nvdaHelper/local/UIAEventLimiter/rateLimitedEventHandler.cpp

```diff
@@ -0,0 +1,227 @@
+/*
+This file is a part of the NVDA project.
+URL: http://github.com/nvaccess/nvda/
+Copyright 2023 NV Access Limited.
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License version 2.0, as published by
+    the Free Software Foundation.
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+This license can be found at:
+http://www.gnu.org/licenses/old-licenses/gpl-2.0.html
+*/
+
+#include <windows.h>
+#include <uiautomation.h>
+#include <numeric>
+#include <variant>
+#include <map>
+#include <vector>
+#include <functional>
+#include <mutex>
+#include <atomic>
+#include <atlcomcli.h>
+#include <comutil.h>
+#include <common/log.h>
+#include "eventRecord.h"
+#include "rateLimitedEventHandler.h"
+
+template<EventRecordConstraints EventRecordClass, typename... EventRecordArgTypes>
+HRESULT RateLimitedEventHandler::queueEvent(EventRecordArgTypes&&... args) {
+	LOG_DEBUG(L"RateLimitedUIAEventHandler::queueEvent called");
+	bool needsFlush = false;
+	const unsigned int flushTimeMS = (EventRecordClass::isCoalesceable)?30:0;
+	{ std::lock_guard lock(mtx);
+		if(!EventRecordClass::isCoalesceable || m_eventRecords.empty()) {
+			needsFlush = true;
+		}
+		LOG_DEBUG(L"RateLimitedUIAEventHandler::queueEvent: Inserting new event");
+		auto& recordVar = m_eventRecords.emplace_back(std::in_place_type_t<EventRecordClass>{}, args...);
+		auto recordVarIter = m_eventRecords.end();
+		recordVarIter--;
+		auto& record = std::get<EventRecordClass>(recordVar);
+		if constexpr(EventRecordClass::isCoalesceable) {
+			LOG_DEBUG(L"RateLimitedUIAEventHandler::queueEvent: Is a coalesceable event");
+			auto coalescingKey = record.generateCoalescingKey();
+			auto existingKeyIter = m_eventRecordsByKey.find(coalescingKey);
+			if(existingKeyIter != m_eventRecordsByKey.end()) {
+				LOG_DEBUG(L"RateLimitedUIAEventHandler::queueEvent: found existing event with same key"); 
+				auto& [existingRecordVarIter,existingCoalesceCount] = existingKeyIter->second;
+				LOG_DEBUG(L"RateLimitedUIAEventHandler::queueEvent: updating key and count to "<<(existingCoalesceCount+1));
+				existingKeyIter->second = {recordVarIter, existingCoalesceCount + 1};
```

Comment:
Are you absolutely certain these vector iterators will point at the same thing always? I would have thought that removing other entries (as can happen below) would invalidate any iterators after that position in the vector such that subsequent iterators would now point to the wrong entry or even be out of bounds.

Reply:
To clarify, they are list iterators, not vector iterators. According to https://en.cppreference.com/w/cpp/container/list
"Adding, removing and moving the elements within the list or across several lists does not invalidate the iterators or references. An iterator is invalidated only when the corresponding element is deleted."

Reply:
Oh, so they are. My bad. Still, I'm glad that's clear up for the record. :)